// zos-lib/lib/utils/ContractsProvider.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _truffleContract = require('truffle-contract');

var _truffleContract2 = _interopRequireDefault(_truffleContract);

var _truffleProvisioner = require('truffle-provisioner');

var _truffleProvisioner2 = _interopRequireDefault(_truffleProvisioner);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
  getFromLib(contractName) {
    // this function should be overwritten when used outside lib
    return this.getByName(contractName);
  },

  getFromKernel(contractName) {
    throw 'getFromKernel function must be implemented';
  },

  getFromArtifacts(name) {
    return artifacts.require(name);
  },

  getByName(name) {
    const path = `${process.cwd()}/build/contracts/${name}.json`;
    return this.getByJSONData(require(path));
  },

  getByJSONData(data) {
    const contract = (0, _truffleContract2.default)(data);
    (0, _truffleProvisioner2.default)(contract, this.artifactsDefaults());
    return contract;
  },

  artifactsDefaults() {
    if (!artifacts) throw "Could not retrieve truffle defaults";
    return artifacts.options || {};
  }
};
