// zos-lib/lib/utils/FileSystem.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.read = read;
exports.readDir = readDir;
exports.exists = exists;
exports.createDir = createDir;
exports.ifExistsThrow = ifExistsThrow;
exports.ifNotExistsThrow = ifNotExistsThrow;
exports.parseJson = parseJson;
exports.parseJsonIfExists = parseJsonIfExists;
exports.editJson = editJson;
exports.writeJson = writeJson;
exports.write = write;
exports.copy = copy;

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function read(filename) {
  return _fs2.default.readFileSync(filename);
}

function readDir(dir) {
  return _fs2.default.readdirSync(dir);
}

function exists(filename) {
  return _fs2.default.existsSync(filename);
}

function createDir(dir) {
  _fs2.default.mkdirSync(dir);
}

function ifExistsThrow(filename, message) {
  if (exists(filename)) throw Error(message);
}

function ifNotExistsThrow(filename, message) {
  if (!exists(filename)) throw Error(message);
}

function parseJson(filename) {
  return JSON.parse(read(filename));
}

function parseJsonIfExists(filename) {
  if (exists(filename)) {
    return JSON.parse(read(filename));
  }
}

function editJson(file, edit) {
  const data = this.parseJson(file);
  edit(data);
  this.writeJson(file, data);
}

function writeJson(filename, data) {
  const json = JSON.stringify(data, null, 2);
  write(filename, json);
}

function write(filename, data) {
  _fs2.default.writeFileSync(filename, data);
}

function copy(source, target) {
  _fs2.default.createReadStream(source).pipe(_fs2.default.createWriteStream(target));
}

exports.default = {
  read,
  readDir,
  exists,
  ifExistsThrow,
  ifNotExistsThrow,
  parseJson,
  createDir,
  editJson,
  parseJsonIfExists,
  writeJson,
  write,
  copy
};
