// zos-lib/lib/release/Release.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _Logger = require('../utils/Logger');

var _Logger2 = _interopRequireDefault(_Logger);

var _ReleaseDeployer = require('./ReleaseDeployer');

var _ReleaseDeployer2 = _interopRequireDefault(_ReleaseDeployer);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const log = new _Logger2.default('Release');

class Release {
  constructor(release, txParams = {}) {
    this._release = release;
    this.txParams = txParams;
  }

  static async deployLocal() {
    return await _ReleaseDeployer2.default.deployLocal(...arguments);
  }

  static async deployDependency() {
    return await _ReleaseDeployer2.default.deployDependency(...arguments);
  }

  address() {
    return this._release.address;
  }

  async owner() {
    return await this._release.owner(this.txParams);
  }

  async freeze() {
    log.info('Freezing release...');
    await this._release.freeze(this.txParams);
  }

  async isFrozen() {
    return await this._release.frozen(this.txParams);
  }

  async getImplementation(contractName) {
    return await this._release.getImplementation(contractName, this.txParams);
  }

  async setImplementation(contractName, implementationAddress) {
    log.info(`Setting ${contractName} implementation ${implementationAddress}`);
    return await this._release.setImplementation(contractName, implementationAddress, this.txParams);
  }
}
exports.default = Release;
